server/controllers/profile.js

```diff
@@ -1,26 +1,171 @@
 const fs = require("fs");
 const User = require("../models/User");
-const asyncHandler = require("express-async-handler");
+const Profile = require("../models/Profile");
 const cloudinary = require("../utils/cloudinaryHelper");
 
+// @route GET /profiles
+// @desc get all profiles
+// @access Public
+exports.loadProfiles = asyncHandler(async (req, res, next) => {
+  const profiles = await Profile.find({});
+
+  if (!profiles.length) {
+    res.status(404);
+    throw new Error("No Profiles found");
+  }
+  res.status(200).send(profiles);
+});
+
+//@route POST /profiles
+//@desc CREATE a new profile
+//@access Private
+exports.createProfile = asyncHandler(async (req, res, next) => {
+  const user = await User.findById(req.user.id);
+  const userId = user?._id;
+  if (!user) {
+    res.status(401);
+    throw new Error("Not authorized");
+  }
+  const {
+    firstName,
+    lastName,
+    address,
+    phone,
+    dob,
+    userPhotoUrl,
+    description,
+    availability,
+    gender,
+  } = req.body;
+
+  const phoneExists = await Profile.findOne({ phone });
+  if (phoneExists) {
+    res.status(400);
+    throw new Error("A User with that phone number already exists");
+  }
+  const profile = await Profile.create({
+    firstName,
+    lastName,
+    dob,
+    phone,
+    address,
+    userPhotoUrl,
+    description,
+    availability,
+    gender,
+    userId,
+  });
+  if (!profile) {
+    res.status(500);
+    throw new Error(
+      "Something went wrong with your profile please try again later"
+    );
+  }
+
+  res.status(201).send(profile);
+});
+
+//@route GET /profiles/:_id
+//@desc find one profile with a particular ID
+//access public
+exports.getProfile = asyncHandler(async (req, res, next) => {
+  const { id } = req.params;
+  const profile = await Profile.findById(id);
+
+  if (!profile) {
+    res.status(404);
+    throw new Error("No Profile found for this user");
+  }
+
+  res.status(200).send(profile);
+});
+
+//@route UPDATE/PUT /profiles/:_id
+//@desc find one profile with a particular ID and update the info within
+//access private
+exports.updateProfile = asyncHandler(async (req, res, next) => {
+  const { id } = req.params;
+  let {
+    firstName,
+    lastName,
+    dob,
+    phone,
+    address,
+    description,
+    availability,
+    gender,
+  } = req.body;
+
+  // Check if the logged in person is the owner of the profile
+  // NOT WORKING rn even though both ids are same
+
+  const user = await User.findById(req.user.id);
+  const profile = await Profile.findById(id);
+  userId = user._id.toString();
+  profileId = profile.userId.toString();
+  if (userId !== profileId) {
+    res.status(404);
+    throw new Error("You are not Authorized to change the data");
+  }
+
+  const newProfile = await Profile.findByIdAndUpdate(id, {
+    firstName,
+    lastName,
+    dob,
+    phone,
+    address,
+    description,
+    availability,
+    gender,
+  });
+  if (!newProfile) {
+    res.status(404);
```

Comment:
I am kinda confused here, should it be 204 as there is no content. Guidance is appreciated.

Reply:
I guess it depends! Is it null here because it was not found or because it didn't successfully update? :)

Reply:
I think in this case its because it didn't successfully update.

Reply:
I also thought this was what you were trying to do since we fetched the profile beforehand (e.g. line `104`), it would make more sense to check whether the profile exists there. 

To keep things simple here, I think we can send a `500`. However, if we wanted to investigate the issue more, we would have to check what caused the update to fail (e.g. validation error, server/db issue...) and send an appropriate status/message for it (e.g. `400`/`409`/`422` or `500`/`503` respectively). 

Keep in mind that more often than not, the recommended approach to updating documents is by using the [`document.save()`](https://mongoosejs.com/docs/documents.html#updating-using-save) function as it will run middleware/validation for us. We can leverage this to identify the nature of the error too with an [error handling middleware](https://mongoosejs.com/docs/middleware.html#error-handling-middleware) (although I would say we would be going beyond the scope of this project at that point).

Reply:
Please update the status code to `500` here :)

Reply:
changed